pkg/controller/job/job_controller.go

```diff
@@ -400,6 +401,83 @@ func (jm *JobController) processNextWorkItem() bool {
 	return true
 }
 
+func mergeResourceChanges(pod *v1.Pod, cMap map[string]*v1.Container) []v1.Container {
+	var resourceUpdates []v1.Container
+
+	// add annotation to pod if request and/or limit from job spec is different
+	for cid, podContainer := range pod.Spec.Containers {
+		hasUpdate := false
+		var patch v1.Container
+		jobSpecContainerResource := cMap[podContainer.Name].Resources
+
+		// update request from job spec
+		if !reflect.DeepEqual(podContainer.Resources.Requests, jobSpecContainerResource.Requests) {
+			if podContainer.Resources.Requests == nil {
+				pod.Spec.Containers[cid].Resources.Requests = make(v1.ResourceList)
+			}
+
+			patch.Resources.Requests = make(v1.ResourceList)
+			for name, jobVal := range jobSpecContainerResource.Requests {
+				if podVal, exists := podContainer.Resources.Requests[name]; !exists ||
+					!reflect.DeepEqual(podVal, jobVal) {
+					patch.Resources.Requests[name] = jobVal
+				}
+			}
+			hasUpdate = true
+		}
+
+		// update limit from job spec
+		if !reflect.DeepEqual(podContainer.Resources.Limits, jobSpecContainerResource.Limits) {
+			if podContainer.Resources.Limits == nil {
+				pod.Spec.Containers[cid].Resources.Limits = make(v1.ResourceList)
+			}
+
+			patch.Resources.Limits = make(v1.ResourceList)
+			for name, jobVal := range jobSpecContainerResource.Limits {
+				if podVal, exists := podContainer.Resources.Limits[name]; !exists ||
+					!reflect.DeepEqual(podVal, jobVal) {
+					patch.Resources.Limits[name] = jobVal
+				}
+			}
+			hasUpdate = true
+		}
+
+		if hasUpdate {
+			patch.Name = podContainer.Name
+			resourceUpdates = append(resourceUpdates, patch)
+		}
+	}
+	return resourceUpdates
+}
+
+func (jm *JobController) patchJobResource(j *batch.Job, pods []*v1.Pod) error {
+	cm := controller.NewPodControllerRefManager(jm.podControl, j, nil, controllerKind, nil)
+
+	cMap := make(map[string]*v1.Container)
+	for i, container := range j.Spec.Template.Spec.Containers {
+		cMap[container.Name] = &(j.Spec.Template.Spec.Containers[i])
+	}
+
+	for _, pod := range pods {
+		resourceUpdates := mergeResourceChanges(pod, cMap)
+		anno := make(map[string]string)
```

Comment:
nitch: better to move the line "anno := make(map[string]string)" into the if block